File: MiP.py
import sys
from work import  returnfulldata, returnuserdata

alldata = returnfulldata()




def matching(userlist, value, TOTALFLEXREQUESTED):
    accepted_offers = {}
    POTENTIALOFFER={}
    BIDS = []

    SIGN= 1
    if TOTALFLEXREQUESTED<0:
        SIGN=-1
        TOTALFLEXREQUESTED = abs(TOTALFLEXREQUESTED)


    for user in userlist:
        userdata, username = returnuserdata(user)
        
        for uservalue in userdata['flexOfferList']:
            if value['RequestID']== uservalue['RequestID']:   
                if (SIGN>0 and uservalue['totalFlexOfferedEU']> 0) or (SIGN<0 and uservalue['totalFlexOfferedEU']<0):
                    if SIGN<0 and uservalue['totalFlexOfferedEU']<0:
                        uservalue['totalFlexOfferedEU'] = abs(uservalue['totalFlexOfferedEU'])
                    BIDS.append((username, uservalue['bidPriceCtpEUList']))


    for name, bidlist in BIDS:
        COUNT=0
        for bids in bidlist:
            if bids<= value['maxPriceCtpEU']:
                POTENTIALOFFER[name+' '+str(COUNT)] = bids
                COUNT+=1

    
    POTENTIALOFFER = {k: v for k, v in sorted(POTENTIALOFFER.items(), key=lambda item: item[1])}

    for i in POTENTIALOFFER:
        name = i.split(' ')[0]
        if TOTALFLEXREQUESTED>1:
            TOTALFLEXREQUESTED-= 1
            if name in accepted_offers:
                accepted_offers[name].append(POTENTIALOFFER[i])
            else:
                accepted_offers[name] = [POTENTIALOFFER[i]]

        else:
            if name in accepted_offers:
                accepted_offers[name].append(POTENTIALOFFER[i])
            else:
                accepted_offers[name] = [POTENTIALOFFER[i]]
            return accepted_offers

    return accepted_offers
    

def checkFulmentFactor(accepted_offers, value,TOTALFLEXREQUESTED):
    SUM = 0

    for i in accepted_offers.values():
        SUM+=len(i)

    if SUM/ abs(TOTALFLEXREQUESTED) * 100 >= value:
        return True

    return False



def mip():
    accepted_offers = {}
    final_accepted_offers = {}

    print('Minimaler Preis ')

    for i in alldata:
        if i['marketType']=='auction':
            for value in i['flexRequestList']:
                if (value['ifFlexRequested'])==False:
                    continue
                print('\n\n\n')

                print('Request for '+value['RequestID']+" "+ str(value['totalFlexRequestedEU']))

                TOTALFLEXREQUESTED = value['totalFlexRequestedEU']

                print('\n')
                userlist = value['loc'].keys()

                accepted_offers = matching(userlist, value, TOTALFLEXREQUESTED)

                if checkFulmentFactor(accepted_offers, value['fullfillmenttFactor'],TOTALFLEXREQUESTED):                
                    final_accepted_offers[value['RequestID']] = accepted_offers
                else:
                    final_accepted_offers[value['RequestID']] = 'fullment factor did not match'

    
    print(final_accepted_offers)
    return final_accepted_offers
    
    

    




if __name__ == '__main__':
    mip()



#   py .\MiP.py


# 1st unit: 7ct, 2nd unit: 8ct, 3rd ubit: 9ct....

File: azure functions/MiP.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FlexchainFunction
{
    public class MiP
    {
        List<FlexibilityRequest> flexRequest;
        List<FlexibilityOffer> flexOffer;

        public MiP(List<FlexibilityRequest> flexRequest, List<FlexibilityOffer> flexOffer)
        {
            this.flexRequest = flexRequest;
            this.flexOffer = flexOffer;
        }


        public Dictionary<string, int> Matching(List<string> userlist, FlexibilityRequest flexibilityRequest, int TOTALFLEXREQUESTED)
        {
            var accepted_offers = new Dictionary<string, int>();
            var POTENTIALOFFER = new Dictionary<string, int>();

            var BIDS = new Dictionary<string, string>();

            var SIGN = 1;

            if (TOTALFLEXREQUESTED < 0)
            {
                SIGN = -1;
                TOTALFLEXREQUESTED = Math.Abs(TOTALFLEXREQUESTED);
            }

            int count = 0;

            foreach (var user in userlist)
            {
                count++;

                for (int j = 0; j < flexOffer.Count; j++)
                {
                    FlexibilityOffer flexibilityOffer = flexOffer[j];

                    if (flexibilityOffer.UserId == user)
                    {
                        foreach (var uservalue in flexibilityOffer.FlexOfferList)
                        {
                            if (uservalue.RequestID == flexibilityRequest.RequestID)
                            {
                                if ((SIGN > 0 && uservalue.totalFlexOfferedEU > 0) || (SIGN < 0 && uservalue.totalFlexOfferedEU < 0))
                                {
                                    if (SIGN < 0 && uservalue.totalFlexOfferedEU < 0)
                                    {
                                        uservalue.totalFlexOfferedEU = Math.Abs(uservalue.totalFlexOfferedEU);

                                    }
                                    BIDS[user] = uservalue.BidPriceCtpEUList;
                                    
                                }
                            }
                        }
                    }
                }
            }

            

            foreach(var username in BIDS.Keys)
            {
                var COUNT = 0;

                var userbids = BIDS[username].Split(", ");

                foreach (var userbid in userbids)
                {
                    if (Int32.Parse(userbid) <= Int32.Parse(flexibilityRequest.MaxPriceCtpEU))
                    {
                        POTENTIALOFFER[ (COUNT).ToString() + " " + username] = Int32.Parse(userbid);

                        COUNT+=1;
                    }
                }

            }

            var sortedDict = POTENTIALOFFER.OrderBy(x => x.Value).ToDictionary(x => x.Key, x => x.Value);


            foreach (var key in sortedDict.Keys)
            {
                

                if(TOTALFLEXREQUESTED>1)
                {
                    TOTALFLEXREQUESTED--;
                    if (accepted_offers.ContainsKey(key))
                    {
                        accepted_offers[key] += POTENTIALOFFER[key];
                    }
                    else
                    {
                        accepted_offers[key] = POTENTIALOFFER[key];
                    }
                }
                else
                {
                    if (accepted_offers.ContainsKey(key))
                    {
                        accepted_offers[key] += POTENTIALOFFER[key];
                    }
                    else
                    {
                        accepted_offers[key] = POTENTIALOFFER[key];
                    }

                    return accepted_offers;
                }
                
            }
            return accepted_offers;

        }


        public bool CheckFulfillmentFactor(Dictionary<string, int> accepted_offers, int fullfillmentFactor, int TOTALFLEXREQUESTED)
        {
            var SUM = 0;

            foreach (KeyValuePair<string, int> pair in accepted_offers)
            {

                SUM+=pair.Value;
            }

            if ((int)(SUM / Math.Abs(TOTALFLEXREQUESTED) * 100) >= fullfillmentFactor)
            {
                return true;
            }
            return false;
        }


        public Flex_MatchingAlgo MiPCalculation()
        {
            var accepted_offers = new Dictionary<string, int>();
            var final_accepted_offers = new Dictionary<string, Dictionary<string, int>>();

            for (int i = 0; i < flexRequest.Count; i++)
            {
                FlexibilityRequest flexibilityRequest = flexRequest[i];

                if (flexibilityRequest.MarketType == "auction")
                {
                    if (flexibilityRequest.IfFlexRequested == false)
                    {
                        var unreachable_accepted_offers = new Dictionary<string, int>();
                        unreachable_accepted_offers.Add("ifflexrequested false", 0);

                        final_accepted_offers[flexibilityRequest.RequestID] = unreachable_accepted_offers;
                    }
                    else
                    {
                        var TOTALFLEXREQUESTED = flexibilityRequest.TotalFlexRequestedEU;
                        List<string> userlist = new List<string>(flexibilityRequest.Loc.Keys);

                        accepted_offers = Matching(userlist, flexibilityRequest, Int32.Parse(TOTALFLEXREQUESTED));

                        

                        if (CheckFulfillmentFactor(accepted_offers, Int32.Parse(flexibilityRequest.FullfillmentFactor), Int32.Parse(TOTALFLEXREQUESTED)))
                        {
                            final_accepted_offers[flexibilityRequest.RequestID] = accepted_offers;
                        }
                        else
                        {
                            var unreachable_accepted_offers = new Dictionary<string, int>();
                            unreachable_accepted_offers.Add("fullfillment factor did not reach", 0);
                            final_accepted_offers[flexibilityRequest.RequestID] = unreachable_accepted_offers;
                        }
                    }
                }
            }

            Flex_MatchingAlgo flex_MatchingAlgo = new Flex_MatchingAlgo(final_accepted_offers);

            return flex_MatchingAlgo;

        }

    }
}
